src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs

```diff
@@ -161,49 +184,152 @@ private static string GetLineInfo(XmlReader reader)
+                element.Attributes.Add(new XmlConfigurationElementAttributeValue(reader.LocalName, reader.Value, GetLineInfo(reader)));
             }
 
             // Go back to the element containing the attributes we just processed
             reader.MoveToElement();
         }
 
         // The special attribute "Name" only contributes to prefix
-        // This method adds a prefix if current node in reader represents a "Name" attribute
-        private static void AddNamePrefix(XmlReader reader, Stack<string> prefixStack,
-            IDictionary<string, string> data, XmlWriter writer)
+        // This method retrieves the Name of the element, if the attribute is present
+        // Unfortunately XmlReader.GetAttribute cannot be used, as it does not support looking for attributes in a case insensitive manner
+        private static string GetName(XmlReader reader)
+        {
+            string name = null;
+
+            while (reader.MoveToNextAttribute())
+            {
+                if (string.Equals(reader.LocalName, NameAttributeKey, StringComparison.OrdinalIgnoreCase))
+                {
+                    // If there is a namespace attached to current attribute
+                    if (!string.IsNullOrEmpty(reader.NamespaceURI))
+                    {
+                        throw new FormatException(SR.Format(SR.Error_NamespaceIsNotSupported, GetLineInfo(reader)));
+                    }
+                    name = reader.Value;
+                    break;
+                }
+            }
+
+            // Go back to the element containing the name we just processed
+            reader.MoveToElement();
+
+            return name;
+        }
+
+        private static IDictionary<string, string> ProvideConfiguration(XmlConfigurationElement? root)
         {
-            if (!string.Equals(reader.LocalName, NameAttributeKey, StringComparison.OrdinalIgnoreCase))
+            var configuration = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            if (root == null)
             {
-                return;
+                return configuration;
             }
 
-            // If current element is not root element
-            if (prefixStack.Any())
+            var rootPrefix = new List<string>();
+
+            // The root element only contributes to the prefix via its Name attribute
+            if (!string.IsNullOrEmpty(root.Name))
             {
-                string lastPrefix = prefixStack.Pop();
-                prefixStack.Push(ConfigurationPath.Combine(lastPrefix, reader.Value));
+                rootPrefix.Add(root.Name);
             }
-            else
+
+            ProcessElementAttributes(rootPrefix, root);
+            ProcessElementContent(rootPrefix, root);
+            ProcessElementChildren(rootPrefix, root);
+
+            return configuration;
+
+            void ProcessElement(List<string> prefix, XmlConfigurationElement element)
```

Comment:
Any reason to make this local methods instead of private static?

Reply:
Two reasons:
1. It allows me to not have to pass around the `Dictionary<string, string> configuration` variable. 
2. These methods conceptually belong to the ProvideConfiguration method, as standalone methods they are not really reusable. They only make sense in the context of the higher level ProvideConfiguration method.

If you prefer private static methods I can change it, obviously.

Reply:
Sounds good. I was curious and missed the fact that you would need to pass the `configuration` dictionary around.